libcudacxx/include/cuda/std/__numeric/saturation_arithmetic.h

```diff
@@ -0,0 +1,802 @@
+        return __clamp_overflow<int8_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 2)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_i16(__x, __y);
+#  elif _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int16_t __result;
+        bool __overflow = _add_overflow_i16(0, __x, __y, &__result);
+        return __clamp_overflow<int16_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 4)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_i32(__x, __y);
+#  elif _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int32_t __result;
+        bool __overflow = _add_overflow_i32(0, __x, __y, &__result);
+        return __clamp_overflow<int32_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 8)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_i64(__x, __y);
+#  elif _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int64_t __result;
+        bool __overflow = _add_overflow_i64(0, __x, __y, &__result);
+        return __clamp_overflow<int64_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+    }
+    else
+    {
+      if constexpr (sizeof(_Tp) == 1)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_u8(__x, __y);
+#  elif _CCCL_COMPILER(MSVC) && _CCCL_ARCH(X86_64)
+        uint8_t __result;
+        bool __overflow = _addcarry_u8(0, __x, __y, &__result);
+        return __clamp_overflow<uint8_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 2)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_u16(__x, __y);
+#  elif _CCCL_COMPILER(MSVC) && _CCCL_ARCH(X86_64)
+        uint16_t __result;
+        bool __overflow = _addcarry_u16(0, __x, __y, &__result);
+        return __clamp_overflow<uint16_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 4)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_u32(__x, __y);
+#  elif _CCCL_COMPILER(MSVC) && _CCCL_ARCH(X86_64)
+        uint32_t __result;
+        bool __overflow = _addcarry_u32(0, __x, __y, &__result);
+        return __clamp_overflow<uint32_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 8)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 41) && _CCCL_ARCH(X86_64)
+        return _sat_add_u64(__x, __y);
+#  elif _CCCL_COMPILER(MSVC) && _CCCL_ARCH(X86_64)
+        uint64_t __result;
+        bool __overflow = _addcarry_u64(0, __x, __y, &__result);
+        return __clamp_overflow<uint64_t>(__x, __y, __result, __overflow);
+#  endif
+      }
+    }
+    return __impl_generic(__x, __y);
```

Comment:
it seems that MSVC messes up the unreachable code analysis here. I believe the issue is that it cannot deal with if constexpr branches that do not return something. 

So we should move this as an `else` branch inside the two branches and add an `_CCCL_UNREACHABLE` at the end